### ISSUE TITLE
Warnings about float to int conversions are wrong

### ISSUE THREAD

**Issue by amaiorano** [microsoft/DirectXShaderCompiler#5961]
See [this comment](https://github.com/microsoft/DirectXShaderCompiler/pull/5805#issuecomment-1772902845) on my PR that has now landed. As shown there, for certain out-of-range float to integral conversions, the compiler warns that the literal value it will use is incorrect. Here's a [DXC Compiler Explorer example](https://godbolt.org/z/9PfEPYa3M) showing these specific warnings, some of which are clearly incorrect.

[Here's a C++ Compiler Explorer example](https://godbolt.org/z/E5nTnx7Kq) showing what Clang 3.8 vs 3.9 does to similar C++ code. Pay attention to the warnings emitted by the compiler, rather than the actual output (the output is mostly garbage, which isn't unexpected since these conversions are UB). We can see that the 3.8 version is what DXC is emitting.

So we should update this warning code with a later version. In fact, in latest Clang, it doesn't bother saying what the literal value will be changed to since it's UB. It says, instead:

```
<source>:28:12: warning: implicit conversion of out of range value from 'double' to 'int' is undefined [-Wliteral-conversion]
```

But for DXC, if we want to make these well-defined, as [my patch](https://github.com/microsoft/DirectXShaderCompiler/pull/5805) did, then we could go with the 3.9 changes instead.

**Comment by amaiorano** [comment]
> @amaiorano is this something you're going to take a look at or should we?

I wasn't planning to work on this any time soon. I also can appreciate that this is not a particularly high priority issue.

**Comment by llvm-beanz** [comment]
All good. We've got it on our backlog. If someone gets bored maybe we'll track down the LLVM 3.8/3.9 change that fixed this and cherry-pick it back.

**Comment by llvm-beanz** [comment]
@amaiorano, this may be resolved with HLSL 202x as a result of https://github.com/microsoft/hlsl-specs/blob/main/proposals/0017-conforming-literals.md.
